### ISSUE TITLE
TLS support for bootstrap server configuration

### ISSUE THREAD

**Issue by ashrivastavagit** [SumoLogic/sumologic-kafka-push#41]
Currently it seems there is support to use TLS for bootstrapServers. We have Kafka brokers where TLS and authentication enabled to connect the broker server. 

Can we have this support, so that we can use this helm chart to publish Kafka logs on sumologic.

**Comment by vsinghal13** [comment]
TLS support is already available. 
The extra volumes example refers to the truststore/keystore mounts needed and the settings just need to be put into the conf identified by endpointsSecret
https://github.com/SumoLogic/sumologic-kafka-push/tree/main/helm#extra-volumes

**Comment by vsinghal13** [comment]
you would include something like:
```
akka: {
  kafka.consumer: {
    security.protocol: SSL
    ssl.truststore.location: /opt/kafka/config/kafka.truststore.jks
    ssl.truststore.password: trustore_password
    ssl.keystore.location: /opt/kafka/config/client.keystore.jks
    ssl.keystore.password: keystore_password
    ssl.key.password: key_password
    ssl.enabled.protocols: TLSv1.2,TLSv1.1,TLSv1
    ssl.client.auth: required
  }
}
```

**Comment by ashrivastavagit** [comment]
Hi Vijit,

Thanks a lot for response.

Do we have option to override “akka” configuration in helm chart, I have not seen this.

If you can share the helm chart actual configuration that will help a lot.

Regards,
Anup

From: Vijit Singhal ***@***.***>
Date: Monday, April 25, 2022 at 8:55 AM
To: SumoLogic/sumologic-kafka-push ***@***.***>
Cc: Anup Kumar Shrivastava ***@***.***>, Author ***@***.***>
Subject: Re: [SumoLogic/sumologic-kafka-push] TLS support for bootstrap server configuration (Issue #41)

you would include something like:

akka: {

  kafka.consumer: {

    security.protocol: SSL

    ssl.truststore.location: /opt/kafka/config/kafka.truststore.jks

    ssl.truststore.password: trustore_password

    ssl.keystore.location: /opt/kafka/config/client.keystore.jks

    ssl.keystore.password: keystore_password

    ssl.key.password: key_password

    ssl.enabled.protocols: TLSv1.2,TLSv1.1,TLSv1

    ssl.client.auth: required

  }

}

**Comment by ashrivastavagit** [comment]
Thanks a lot Vijit,

I will work on it and will update on this thread for sure.

Regards,
Anup

From: Vijit Singhal ***@***.***>
Date: Monday, April 25, 2022 at 9:31 AM
To: SumoLogic/sumologic-kafka-push ***@***.***>
Cc: Anup Kumar Shrivastava ***@***.***>, Author ***@***.***>
Subject: Re: [SumoLogic/sumologic-kafka-push] TLS support for bootstrap server configuration (Issue #41)

you can take a look at: https://github.com/SumoLogic/sumologic-kafka-push/blob/main/helm/README.md#configuration

Basically, the additional conf needs to be put in a file and that file will be specified at endpointsSecret parameter in the helm chart.

**Comment by vsinghal13** [comment]
were you able to run `helm upgrade` successfully after the changes that you made?

**Comment by ddaughertysl** [comment]
On pod startup you should see in the logs the kafka consumer configuration. Does it match with what you are setting in the endpointsSecret? The secret should also be formatted as follows: 

{
   "application.conf": "{\"akka\":{\"kafka.consumer\": { ... } }"
}

**Comment by vsinghal13** [comment]
Please have a look at https://github.com/SumoLogic/sumologic-kafka-push/pull/45

**Comment by aporwal3** [comment]
@vsinghal13 Have applied the same configuration as what you have mentioned in your PR. But still the SSL configuration dont works in our environment. We have other applications as well on our cluster which uses the same SSL configuration and it does work on them. 

Also did performed the troubleshooting steps as well and all looks correct from that front.

So can you please advise as what might be going wrong ?

Logs:
`2022-06-24 11:28:34 INFO  [.default-dispatcher-3]  a.e.s.Slf4jLogger - Slf4jLogger started
2022-06-24 11:28:35 INFO  [.default-dispatcher-7]  a.k.i.SingleSourceLogic - [066ce] Starting. StageActor Actor[akka://sumo-push-main/system/Materializers/StreamSupervisor-0/$$c#97301754]
2022-06-24 11:28:35 INFO  [default-dispatcher-10]  o.a.k.c.c.ConsumerConfig - ConsumerConfig values: 
	allow.auto.create.topics = true
	auto.commit.interval.ms = 5000
	auto.offset.reset = earliest
	bootstrap.servers = [10.2x0.x.27:9093, 10.2x0.x.40:9093, 10.2x0.x.137:9093, 10.2x0.x.178:9093]
	check.crcs = true
	client.dns.lookup = use_all_dns_ips
	client.id = consumer-kafka-push-logs-1
	client.rack = 
	connections.max.idle.ms = 540000
	default.api.timeout.ms = 60000
	enable.auto.commit = false
	exclude.internal.topics = true
	fetch.max.bytes = 52428800
	fetch.max.wait.ms = 500
	fetch.min.bytes = 1
	group.id = kafka-push-logs
	group.instance.id = null
	heartbeat.interval.ms = 3000
	interceptor.classes = []
	internal.leave.group.on.close = true
	internal.throw.on.fetch.stable.offset.unsupported = false
	isolation.level = read_uncommitted
	key.deserializer = class org.apache.kafka.common.serialization.StringDeserializer
	max.partition.fetch.bytes = 1048576
	max.poll.interval.ms = 300000
	max.poll.records = 500
	metadata.max.age.ms = 300000
	metric.reporters = []
	metrics.num.samples = 2
	metrics.recording.level = INFO
	metrics.sample.window.ms = 30000
	partition.assignment.strategy = [class org.apache.kafka.clients.consumer.RangeAssignor]
	receive.buffer.bytes = 65536
	reconnect.backoff.max.ms = 1000
	reconnect.backoff.ms = 50
	request.timeout.ms = 30000
	retry.backoff.ms = 100
	sasl.client.callback.handler.class = null
	sasl.jaas.config = [hidden]
	sasl.kerberos.kinit.cmd = /usr/bin/kinit
	sasl.kerberos.min.time.before.relogin = 60000
	sasl.kerberos.service.name = null
	sasl.kerberos.ticket.renew.jitter = 0.05
	sasl.kerberos.ticket.renew.window.factor = 0.8
	sasl.login.callback.handler.class = null
	sasl.login.class = null
	sasl.login.refresh.buffer.seconds = 300
	sasl.login.refresh.min.period.seconds = 60
	sasl.login.refresh.window.factor = 0.8
	sasl.login.refresh.window.jitter = 0.05
	sasl.mechanism = GSSAPI
	security.protocol = SSL
	security.providers = null
	send.buffer.bytes = 131072
	session.timeout.ms = 10000
	socket.connection.setup.timeout.max.ms = 127000
	socket.connection.setup.timeout.ms = 10000
	ssl.cipher.suites = null
	ssl.enabled.protocols = [TLSv1.2, TLSv1.1, TLSv1]
	ssl.endpoint.identification.algorithm = https
	ssl.engine.factory.class = null
	ssl.key.password = [hidden]
	ssl.keymanager.algorithm = SunX509
	ssl.keystore.certificate.chain = null
	ssl.keystore.key = null
	ssl.keystore.location = /etc/ssl/certs/kafka/truststore.jks
	ssl.keystore.password = [hidden]
	ssl.keystore.type = JKS
	ssl.protocol = TLSv1.3
	ssl.provider = null
	ssl.secure.random.implementation = null
	ssl.trustmanager.algorithm = PKIX
	ssl.truststore.certificates = null
	ssl.truststore.location = /etc/ssl/certs/kafka/truststore.jks
	ssl.truststore.password = [hidden]
	ssl.truststore.type = JKS
	value.deserializer = class com.sumologic.sumopush.serde.KubernetesLogEventSerde$

2022-06-24 11:28:35 WARN  [default-dispatcher-10]  o.a.k.c.c.ConsumerConfig - The configuration 'ssl.client.auth' was supplied but isn't a known config.
2022-06-24 11:28:35 INFO  [default-dispatcher-10]  o.a.k.c.u.AppInfoParser - Kafka version: 2.7.0
2022-06-24 11:28:35 INFO  [default-dispatcher-10]  o.a.k.c.u.AppInfoParser - Kafka commitId: 448719dc99a19793
2022-06-24 11:28:35 INFO  [default-dispatcher-10]  o.a.k.c.u.AppInfoParser - Kafka startTimeMs: 1656070115704
2022-06-24 11:28:35 INFO  [default-dispatcher-10]  o.a.k.c.c.KafkaConsumer - [Consumer clientId=consumer-kafka-push-logs-1, groupId=kafka-push-logs] Subscribed to pattern: 'v2dev-instaclustr.*-logs'
2022-06-24 11:28:36 WARN  [default-dispatcher-24]  o.a.k.c.NetworkClient - [Consumer clientId=consumer-kafka-push-logs-1, groupId=kafka-push-logs] Bootstrap broker 10.220.5.27:9093 (id: -1 rack: null) disconnected
2022-06-24 11:28:37 WARN  [default-dispatcher-25]  o.a.k.c.NetworkClient - [Consumer clientId=consumer-kafka-push-logs-1, groupId=kafka-push-logs] Bootstrap broker 10.220.5.100:9093 (id: -3 rack: null) disconnected
2022-06-24 11:28:37 WARN  [default-dispatcher-10]  o.a.k.c.NetworkClient - [Consumer clientId=consumer-kafka-push-logs-1, groupId=kafka-push-logs] Bootstrap broker 10.220.5.92:9093 (id: -4 rack: null) disconnected
`

**Comment by vsinghal13** [comment]
@ddaughertysl any suggestions on this?

**Comment by ddaughertysl** [comment]
i think this shows progress in that the ssl settings are now being applied and the security protocol is SSL. The question is why aren't they working. Which is a bit harder for us to answer without comparing both the settings that work in your other applications and the sumo-push settings.  Also i notice that your ssl.protocol is set to TLSv1.3 and ssl.enabled.protocols don't include that value.  So you might also try including v1.3 in the list of enabled protocols.

**Comment by aporwal3** [comment]
@ddaughertysl Did the requested change but still the issue persist.

2022-07-01 09:27:03 INFO  [.default-dispatcher-3]  a.e.s.Slf4jLogger - Slf4jLogger started
2022-07-01 09:27:05 INFO  [.default-dispatcher-5]  a.k.i.SingleSourceLogic - [f1e6a] Starting. StageActor Actor[akka://sumo-push-main/system/Materializers/StreamSupervisor-0/$$c#-38628872]
2022-07-01 09:27:05 INFO  [default-dispatcher-10]  o.a.k.c.c.ConsumerConfig - ConsumerConfig values: 
	allow.auto.create.topics = true
	auto.commit.interval.ms = 5000
	auto.offset.reset = earliest
	bootstrap.servers = [10.x.27:9093, 10.x.40:9093, 10.x.100:9093, 10.x.92:9093, 10.x.137:9093, 10.x.178:9093]
	check.crcs = true
	client.dns.lookup = use_all_dns_ips
	client.id = consumer-kafka-push-logs-1
	client.rack = 
	connections.max.idle.ms = 540000
	default.api.timeout.ms = 60000
	enable.auto.commit = false
	exclude.internal.topics = true
	fetch.max.bytes = 52428800
	fetch.max.wait.ms = 500
	fetch.min.bytes = 1
	group.id = kafka-push-logs
	group.instance.id = null
	heartbeat.interval.ms = 3000
	interceptor.classes = []
	internal.leave.group.on.close = true
	internal.throw.on.fetch.stable.offset.unsupported = false
	isolation.level = read_uncommitted
	key.deserializer = class org.apache.kafka.common.serialization.StringDeserializer
	max.partition.fetch.bytes = 1048576
	max.poll.interval.ms = 300000
	max.poll.records = 500
	metadata.max.age.ms = 300000
	metric.reporters = []
	metrics.num.samples = 2
	metrics.recording.level = INFO
	metrics.sample.window.ms = 30000
	partition.assignment.strategy = [class org.apache.kafka.clients.consumer.RangeAssignor]
	receive.buffer.bytes = 65536
	reconnect.backoff.max.ms = 1000
	reconnect.backoff.ms = 50
	request.timeout.ms = 30000
	retry.backoff.ms = 100
	sasl.client.callback.handler.class = null
	sasl.jaas.config = [hidden]
	sasl.kerberos.kinit.cmd = /usr/bin/kinit
	sasl.kerberos.min.time.before.relogin = 60000
	sasl.kerberos.service.name = null
	sasl.kerberos.ticket.renew.jitter = 0.05
	sasl.kerberos.ticket.renew.window.factor = 0.8
	sasl.login.callback.handler.class = null
	sasl.login.class = null
	sasl.login.refresh.buffer.seconds = 300
	sasl.login.refresh.min.period.seconds = 60
	sasl.login.refresh.window.factor = 0.8
	sasl.login.refresh.window.jitter = 0.05
	sasl.mechanism = GSSAPI
	security.protocol = SSL
	security.providers = null
	send.buffer.bytes = 131072
	session.timeout.ms = 10000
	socket.connection.setup.timeout.max.ms = 127000
	socket.connection.setup.timeout.ms = 10000
	ssl.cipher.suites = null
	ssl.enabled.protocols = [TLSv1.3, TLSv1.2, TLSv1.1, TLSv1]
	ssl.endpoint.identification.algorithm = https
	ssl.engine.factory.class = null
	ssl.key.password = [hidden]
	ssl.keymanager.algorithm = SunX509
	ssl.keystore.certificate.chain = null
	ssl.keystore.key = null
	ssl.keystore.location = /etc/ssl/certs/kafka/truststore.jks
	ssl.keystore.password = [hidden]
	ssl.keystore.type = JKS
	ssl.protocol = TLSv1.3
	ssl.provider = null
	ssl.secure.random.implementation = null
	ssl.trustmanager.algorithm = PKIX
	ssl.truststore.certificates = null
	ssl.truststore.location = /etc/ssl/certs/kafka/truststore.jks
	ssl.truststore.password = [hidden]
	ssl.truststore.type = JKS
	value.deserializer = class com.sumologic.sumopush.serde.KubernetesLogEventSerde$

2022-07-01 09:27:05 WARN  [default-dispatcher-10]  o.a.k.c.c.ConsumerConfig - The configuration 'ssl.client.auth' was supplied but isn't a known config.
2022-07-01 09:27:05 INFO  [default-dispatcher-10]  o.a.k.c.u.AppInfoParser - Kafka version: 2.7.0
2022-07-01 09:27:05 INFO  [default-dispatcher-10]  o.a.k.c.u.AppInfoParser - Kafka commitId: 448719dc99a19793
2022-07-01 09:27:05 INFO  [default-dispatcher-10]  o.a.k.c.u.AppInfoParser - Kafka startTimeMs: 1656667625750
2022-07-01 09:27:05 INFO  [default-dispatcher-10]  o.a.k.c.c.KafkaConsumer - [Consumer clientId=consumer-kafka-push-logs-1, groupId=kafka-push-logs] Subscribed to pattern: 'v2dev-instaclustr.*-logs'
2022-07-01 09:27:06 WARN  [default-dispatcher-24]  o.a.k.c.NetworkClient - [Consumer clientId=consumer-kafka-push-logs-1, groupId=kafka-push-logs] Bootstrap broker 10.x.27:9093 (id: -1 rack: null) disconnected
2022-07-01 09:27:07 WARN  [default-dispatcher-23]  o.a.k.c.NetworkClient - [Consumer clientId=consumer-kafka-push-logs-1, groupId=kafka-push-logs] Bootstrap broker 10.x.100:9093 (id: -3 rack: null) disconnected
2022-07-01 09:27:07 WARN  [default-dispatcher-24]  o.a.k.c.NetworkClient - [Consumer clientId=consumer-kafka-push-logs-1, groupId=kafka-push-logs] Bootstrap broker 10.x.137:9093 (id: -5 rack: null) disconnected
2022-07-01 09:27:08 WARN  [default-dispatcher-23]  o.a.k.c.NetworkClient - [Consumer clientId=consumer-kafka-push-logs-1, groupId=kafka-push-logs] Bootstrap broker 10.x.40:9093 (id: -2 rack: null) disconnected

**Comment by ddaughertysl** [comment]
have you configured your `security.protocol` and `sasl.mechanism` as follows? https://kafka.apache.org/documentation/#security_sasl_scram_clientconfig

**Comment by aporwal3** [comment]
@ddaughertysl thanks again. it has worked now. But log parsing is failing now with below error. Do we need to configure serdeClass ?

Error Logs:

2022-07-06 18:19:16 ERROR [default-dispatcher-28]  c.s.s.a.LogProcessor$ - unable to parse log message payload: Jul 01 13:42:06 8f844057-a8f5-4f0d-9b66-3347d9815750 [2022-07-01 13:42:06,957] INFO [SocketServer listenerType=ZK_BROKER, nodeId=2] Failed authentication with /10.3.129.8 (Unexpected Kafka request of type METADATA during SASL handshake.) (org.apache.kafka.common.network.Selector)
java.lang.Exception: payload: Jul 01 13:42:06 8f844057-a8f5-4f0d-9b66-3347d9815750 [2022-07-01 13:42:06,957] INFO [SocketServer listenerType=ZK_BROKER, nodeId=2] Failed authentication with /10.3.129.8 (Unexpected Kafka request of type METADATA during SASL handshake.) (org.apache.kafka.common.network.Selector)
	at com.sumologic.sumopush.serde.KubernetesLogEventSerde$.deserialize(KubernetesLogEventSerde.scala:15)
	at com.sumologic.sumopush.serde.KubernetesLogEventSerde$.deserialize(KubernetesLogEventSerde.scala:8)
	at org.apache.kafka.common.serialization.Deserializer.deserialize(Deserializer.java:60)
	at com.sumologic.sumopush.serde.LogEventSerde.deserialize(LogEventSerde.scala:8)
	at org.apache.kafka.clients.consumer.internals.Fetcher.parseRecord(Fetcher.java:1387)
	at org.apache.kafka.clients.consumer.internals.Fetcher.access$3400(Fetcher.java:133)
	at org.apache.kafka.clients.consumer.internals.Fetcher$CompletedFetch.fetchRecords(Fetcher.java:1618)
	at org.apache.kafka.clients.consumer.internals.Fetcher$CompletedFetch.access$1700(Fetcher.java:1454)
	at org.apache.kafka.clients.consumer.internals.Fetcher.fetchRecords(Fetcher.java:687)
	at org.apache.kafka.clients.consumer.internals.Fetcher.fetchedRecords(Fetcher.java:638)
	at org.apache.kafka.clients.consumer.KafkaConsumer.pollForFetches(KafkaConsumer.java:1272)
	at org.apache.kafka.clients.consumer.KafkaConsumer.poll(KafkaConsumer.java:1233)
	at org.apache.kafka.clients.consumer.KafkaConsumer.poll(KafkaConsumer.java:1206)
	at akka.kafka.internal.KafkaConsumerActor.poll(KafkaConsumerActor.scala:549)
	at akka.kafka.internal.KafkaConsumerActor$$anonfun$regularReceive$1.applyOrElse(KafkaConsumerActor.scala:296)
	at akka.actor.Actor.aroundReceive(Actor.scala:537)
	at akka.actor.Actor.aroundReceive$(Actor.scala:535)
	at akka.kafka.internal.KafkaConsumerActor.akka$actor$Timers$$super$aroundReceive(KafkaConsumerActor.scala:205)
	at akka.actor.Timers.aroundReceive(Timers.scala:56)
	at akka.actor.Timers.aroundReceive$(Timers.scala:41)
	at akka.kafka.internal.KafkaConsumerActor.aroundReceive(KafkaConsumerActor.scala:205)
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:577)
	at akka.actor.ActorCell.invoke(ActorCell.scala:547)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:270)
	at akka.dispatch.Mailbox.run(Mailbox.scala:231)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: org.json4s.ParserUtil$ParseException: unknown token J
Near: J
	at org.json4s.native.JsonParser$Parser.fail(JsonParser.scala:236)
	at org.json4s.native.JsonParser$Parser.nextToken(JsonParser.scala:324)
	at org.json4s.native.JsonParser$.$anonfun$astParser$1(JsonParser.scala:188)
	at org.json4s.native.JsonParser$.$anonfun$astParser$1$adapted(JsonParser.scala:145)
	at org.json4s.native.JsonParser$.parse(JsonParser.scala:133)
	at org.json4s.native.JsonParser$.parse(JsonParser.scala:71)
	at org.json4s.native.JsonParser$.parse(JsonParser.scala:50)
	at org.json4s.native.Serialization$.read(Serialization.scala:71)
	at org.json4s.Serialization.read(Serialization.scala:25)
	at org.json4s.Serialization.read$(Serialization.scala:25)
	at org.json4s.native.Serialization$.read(Serialization.scala:32)
	at com.sumologic.sumopush.model.KubernetesLogEventSerializer$.fromJson(KubernetesLogEvent.scala:47)
	at com.sumologic.sumopush.serde.KubernetesLogEventSerde$.deserialize(KubernetesLogEventSerde.scala:13)
	... 27 common frames omitted

**Comment by ddaughertysl** [comment]
Yes you should be using the JsonLogEventSerde but it also looks like your messages aren't in Json format. Can you fix that?

**Comment by ddaughertysl** [comment]
Yes per the docs https://github.com/SumoLogic/sumologic-kafka-push#supported-message-formats we only support json format at this time.

**Comment by ashrivastavagit** [comment]
Yes. I had to adjust resource configuration.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
